### PR TITLE
[LSP] Add semantic token caching for solution open

### PR DESCRIPTION
**Pull request by allisonchou**
- In local C#/VB, semantic tokens are cached in a SQL database that is queried on solution open. This allows us to instantly colorize the solution without performing any token computations. The goal of this PR is to accomplish something similar in LSP scenarios.
- This PR adds the caching described above to Roslyn LSP. It also makes it so we now only return semantic classifications, whereas before we returned both semantic + syntactic classifications. We are able to do this because the C# syntactic classifier runs on the client, so we save ourselves lots of work by only returning semantic classifications.
- We have to special case Razor since Razor requires us to return both syntactic and semantic classifications. This is due to the C# syntactic classifier not running on the client in Razor scenarios. Ideally, Razor would eventually run the classifier in some way (tracked by https://github.com/dotnet/razor-tooling/issues/5850) so we can get rid of this special casing. For now, we cache and return both types of tokens.
- Updated + added tests.

### CODE REVIEW

File: src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationUtilities.cs
Comment (allisonchou) [inline src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationUtilities.cs:152]: The `isRazorDoc` parameter was added for testing purposes so we can compare returning only semantic tokens vs. returning both semantic and syntactic tokens.

File: src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationUtilities.cs
Comment (allisonchou) [inline src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationUtilities.cs:169]: This method and the method below were moved to `SemanticClassificationCacheUtilities.cs` so that they're accessible in the LSP layer.

File: src/Features/Core/Portable/SemanticClassificationCache/SemanticClassificationCacheUtilities.cs
Comment (allisonchou) [inline src/Features/Core/Portable/SemanticClassificationCache/SemanticClassificationCacheUtilities.cs:63]: This if-check and the added `isRazorDoc` parameter are the only new parts of these methods; they were otherwise moved untouched from `SemanticClassificationUtilities.cs`. As previously mentioned, the `isRazorDoc` parameter was added for testing purposes.

File: src/Workspaces/Remote/ServiceHub/Services/SemanticClassificationCache/RemoteSemanticClassificationCacheService.cs
Comment (allisonchou) [inline src/Workspaces/Remote/ServiceHub/Services/SemanticClassificationCache/RemoteSemanticClassificationCacheService.cs:106]: `.IsRazorDocument()` doesn't work in OOP scenarios, so as a workaround I added a check for `.EndsWith(".g.cs")` instead which seems to do the trick

File: src/Features/Core/Portable/SemanticClassificationCache/SemanticClassificationCacheUtilities.cs
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SemanticClassificationCache/SemanticClassificationCacheUtilities.cs:72]: hrmm... should this code just move up to the razor caller instead?

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:189]: so instead of passing isRazorDoc here (and then having AddSemandticClassificationsAsync do odd stuff (like adding syntactic classifications), i would instead have the code *here* fork it's behavior (with the good comment you have), and say htat in the razor case we get all diagnostics, where for all otehrs we just get semantic classification.

File: src/VisualStudio/Core/Def/Implementation/SemanticClassificationCache/SemanticClassificationCacheIncrementalAnalyzerProvider.cs
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/SemanticClassificationCache/SemanticClassificationCacheIncrementalAnalyzerProvider.cs:66]: @dibarbet here as well...

interesting.  why is this @allisonchou ?  why not have the normal incremental analyzer handle this.  note: if this is the right way to go, then this comment needs beefing up.  Specifically, this is an example of a comment htat just repeats what the code is doing but doesn't provide clarity as to 'why'.  I cannot look at this code/comment and say: ah, yes, that makes sense, we def should be doing things in this fashion.

File: src/Workspaces/Remote/ServiceHub/Services/SemanticClassificationCache/RemoteSemanticClassificationCacheService.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Services/SemanticClassificationCache/RemoteSemanticClassificationCacheService.cs:106]: @NTaylorMullen for thoughts.

Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Services/SemanticClassificationCache/RemoteSemanticClassificationCacheService.cs:106]: note: This feels weird as well.  somethign about the entire razor space here seems somewhat broken (necessitating these extra strange codepaths at the feature level).  my preference would be that if razor is *currently* doing something different/weird, that we just not have this optimization/feature for them.  and, if we can then get razor docs/scenarios to work like normal scenarios, then we reenable, with no specialized code *in the feature* to handle razor.

File: src/VisualStudio/Core/Def/Implementation/SemanticClassificationCache/SemanticClassificationCacheIncrementalAnalyzerProvider.cs
Comment (allisonchou) [reply src/VisualStudio/Core/Def/Implementation/SemanticClassificationCache/SemanticClassificationCacheIncrementalAnalyzerProvider.cs:66]: The main issue with using the default logic is the method's call to [`document.IsOpen()`](https://github.com/dotnet/roslyn/blob/498d5b1f6feace039534035a08aa864c0cd26d1e/src/VisualStudio/Core/Def/Implementation/SemanticClassificationCache/SemanticClassificationCacheIncrementalAnalyzerProvider.cs#L48). This call is problematic in LSP for two reasons:

1) It bases whether the document is open on the non-LSP workspace rather than the LSP workspace.
2) Razor back-end generated documents won't be considered open even if the front-end document is open. This means we would have to special case Razor (via exposing a new document service, workspace service or something else).

Putting the caching logic in the LSP handler addresses both 1) and 2) by ensuring we only cache open files in the LSP workspace, since we only get semantic token requests for the file(s) currently open. In an ideal world, I think we'd have the `IsOpen` check correspond to the LSP workspace and that workspace can then specially check for open Razor documents, but I believe that is a much bigger work item.

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs:88]: (Same explanation as https://github.com/dotnet/roslyn/pull/58238#discussion_r768215940)

File: src/Workspaces/Remote/ServiceHub/Services/SemanticClassificationCache/RemoteSemanticClassificationCacheService.cs
Comment (allisonchou) [reply src/Workspaces/Remote/ServiceHub/Services/SemanticClassificationCache/RemoteSemanticClassificationCacheService.cs:106]: Most of this special casing is due to Razor not running the C# syntax classifier on their end. However, I believe adding support is a pretty big work item and could possibly also negatively impact Razor perf, so the effects would have to be investigated.

While this special casing doesn't look good, I think it will result in pretty good gains for Razor semantic tokens which imo might be an appropriate trade-off since the Razor LSP editor is on by default for users.

**Comment by allisonchou** [comment]
@CyrusNajmabadi whenever you have time, this should be ready to review again. I've addressed your feedback + responded to the changes from your other PRs.

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:198]: nice.  this is all much cleaner IMO :)

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:415]: note: as per previous discussion, i don't think this is necessary.  but if you want to keep it, that's fine.

File: src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:119]: i don't love this.  primarily because you're overiwring a variable that feels like it is controlled by the using block.  Technically, this works... it's just oogy :)

File: src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:199]: why this change?

File: src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:215]: thanks for the example.  do we have ests for this?

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:187]: so, i have a subjective question for you.   It was def me that pushed using ClassifierHelper.getClassifiedSpansAsync.  However, in order to do so you had to add a fair amount of logic to it to get what you need for the LSP case.

My overall question is which do *you* think is better of the following two options:

1. keep the PR as is, and have the extra complexity in that method.
2. just provide a totally different method that, while similar to the above, is independent and dedicated for the LSP case.

If you feel like '2' actually ends up adding/repeating too much code.  Then let's stick with what's in the PR. However, if you feel that the two methods are different enough that trying to force them to be one method (along with several booleans controlling) is just to overwrought, then maybe we should just have two.

I don't have enough context to say from reading this.  However, my hope was that you could just use hte method as is.  the need to pass all this customization logic in is making me rethink that :)

Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:187]: Just tested out option 2). Unfortunately, it does feel like a substantial amount of the code is duplicated between the LSP and non-LSP methods. The only portions that differ are a) whether to remove additive spans and b) whether to fill in classified span gaps. The optional parameters don't feel great to me either, but I feel like the amount of duplication we avoid may be worth the tradeoff.

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:415]: Interestingly enough, `GetClassifiedSpansAsync` and `AddSemanticClassificationsAsync` each return classifications in opposite order from one another. I think we might want to keep the check just in case.

File: src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs
Comment (allisonchou) [reply src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs]: Gotcha. I slightly refactored this portion in my latest push just now, to me it looks a little cleaner but let me know if you still have any concerns.

File: src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs
Comment (allisonchou) [reply src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:213]: Yep, a test in `SemanticTokensRangeTests` actually caught this error 😄

File: src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs
Comment (allisonchou) [reply src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:197]: I believe this is an existing bug in the logic related to the comment I added below:

```
                    // There may be multiple semantic parts corresponding to a single
                    // syntactic part, so we might need to go through a syntactic part
                    // multiple times to verify. For example, this is the case with
                    // verbatim string literals containing string escape characters.
```

Essentially, there's a chance the TextSpan of the semantic part will not exactly match the TextSpan of the syntactic part. This occurs with string literals containing string escape characters. The string escape character will not have the same exact span as the syntactic part and therefore will not be considered. I believe instead calling `OverlapsWith` mitigates the problem.

**Comment by allisonchou** [comment]
@CyrusNajmabadi thanks for your review! I've responded to your feedback. Would you mind taking a look again whenever you have time?

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:169]: more of a naming nit - perhaps we could rename this to be something like `includeSyntacticClassifications`

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs
Comment (ryanbrandenburg) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:154]: Not a fan of this kind of "secret return" inside of the `classifiedSpans`. Could `GetClassifiedSpansForDocumentAsync` return `Task<ClassifiedSpan[]>` so we can be explicit about what had work done on it? Also, if `ArrayBuilder<ClassifiedSpan>` is static couldn't we avoid passing it and just create the instance inside of `GetClassifiedSpansForDocumentAsync`?

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs
Comment (ryanbrandenburg) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs:53]: Nit: Personally I'm of the opinion that parameters should be all on the same line or each on their own line to avoid diffs like this.

File: src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensRangeTests.cs
Comment (ryanbrandenburg) [inline src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensRangeTests.cs:22]: Why change the name? Seems like that'd mess up statistics if nothing else.

**Review by ryanbrandenburg — COMMENTED** [review]
Only complaint is the "secret return", but you probably want to wait for review from someone with more expertise in Roslyn.

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:186]: might be good to make a note of why the flags are set this way for LSP.  I saw a mention below for removeAdditiveSpans, but not for fillInClassifiedSpanGaps

File: src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensRangeTests.cs
Comment (dibarbet) [inline src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensRangeTests.cs:50]: did this test go missing?  I'm not seeing a new test that verifies the partial (1,0) -> (2,0) positions

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs
Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:154]: note: in terms of naming, we try to do 'Get' to have a return value, and 'Add' if someone will pass in the collection to fill.

File: src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:117]: this is *definitely* not going to work.  :)  remember that when the using goes out of scope, we'll put the array back in teh pool.  meaning someone else might grab on it and party on it at the same time as us.  i have a recommendtation on how to structure this code.  ping me :)

File: src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:213]: ok great!

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:213]: actually, can you ensure we have a few tests here.  for example, if a verbatim literal contains multiple escape characters like so:

`$"foo""bar""baz{x + y}a""b""c"`

something to stress teh system

File: src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs
Comment (allisonchou) [reply src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:119]: Ok, I actually had to revert my refactoring - guess the logic wasn't equivalent 😦 lmk if you have any other suggestions

File: src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs
Comment (allisonchou) [reply src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:212]: I added a regex test in my latest push which I believe accomplishes this 🙂

File: src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensRangeTests.cs
Comment (allisonchou) [reply src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensRangeTests.cs:50]: Whoops, you're totally right. Thanks! Added it back.

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs
Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:190]: sgtm.

**Review by CyrusNajmabadi — APPROVED** [review]
Thanks!

**Review by dibarbet — APPROVED** [review]

**Comment by allisonchou** [comment]
Thanks everyone!

**Pull request merged**
